src/components/ReportsPage/HeatmapPage/HeatmapUtils.ts

```diff
@@ -119,14 +113,46 @@ export const getScorecardServiceScoresByGroupByOption = (
   }
 };
 
+interface ScorecardRuleMetadata {
+  ruleName: string;
+  totalServicesScore: number;
+  numApplicableServices: number;
+}
+
 export const getAverageRuleScores = (
   scores: ScorecardServiceScore[],
-  serviceCount: number,
 ): number[] => {
-  return scores
-    .map(score => getSortedRulesFromScores(score).map(rule => rule.score))
-    .reduce((r, a) => a.map((b, i) => (r[i] || 0) + (b ? 1 : 0)), [])
-    .map(score => score / serviceCount);
+  const ruleIdToScoreMetadata: Record<number, ScorecardRuleMetadata> = {};
+  scores.forEach(score =>
+    score.rules.forEach(ruleOutcome => {
+      if (ruleIdToScoreMetadata[ruleOutcome.rule.id] === undefined) {
+        ruleIdToScoreMetadata[ruleOutcome.rule.id] = {
+          ruleName: ruleName(ruleOutcome.rule),
+          totalServicesScore: 0,
+          numApplicableServices: 0,
+        };
+      }
+      if (isApplicableRuleOutcome(ruleOutcome)) {
+        const scorecardRuleMetadata =
+          ruleIdToScoreMetadata[ruleOutcome.rule.id];
+        ruleIdToScoreMetadata[ruleOutcome.rule.id] = {
+          ruleName: ruleName(ruleOutcome.rule),
+          totalServicesScore:
+            scorecardRuleMetadata.totalServicesScore +
+            (ruleOutcome.score ? 1 : 0),
+          numApplicableServices:
+            scorecardRuleMetadata.numApplicableServices + 1,
+        };
+      }
+    }),
+  );
+  return values(ruleIdToScoreMetadata)
```

Comment:
[Non-blocking] Why are we creating a dictionary just to return it to a list? Can there be duplicate rule ids in the scores list, and thus we are using it as a way to deduplicate?

Reply:
I don't think there can be duplicates. I used a dictionary so that we don't have to iterate through the entire list to look for the relevant rule to update calculations for.